cvat/apps/engine/static/engine/js/player.js

```diff
@@ -14,121 +14,228 @@
+                this._requestedChunks[chunkIdx].buffer[last] = frameData;
+                this._requestedChunks[chunkIdx].requestedFrames.delete(last);
+                if (this._requestedChunks[chunkIdx].requestedFrames.size === 0) {
+                    if (this._requestedChunks[chunkIdx].resolve) {
+                        const bufferedframes = Object.keys(
+                            this._requestedChunks[chunkIdx].buffer,
+                        ).map(f => +f);
+                        this._requestedChunks[chunkIdx].resolve(new Set(bufferedframes));
+                    }
                 }
             }
+        } else if (!isReject) {
+            this._loaded = last;
+            this.notify();
         }
     }
 
-    _load() {
-        if (!this._loadInterval) {
-            this._loadInterval = setInterval(() => {
-                if (!this._loadAllowed) {
-                    return;
-                }
+    requestOneChunkFrames(chunkIdx) {
+        return new Promise((resolve, reject) => {
+            this._requestedChunks[chunkIdx] = {
+                ...this._requestedChunks[chunkIdx],
+                resolve,
+                reject,
+            };
+            for (const frame of this._requestedChunks[chunkIdx].requestedFrames.entries()) {
+                const requestedFrame = frame[1];
+                this._frameProvider.require(requestedFrame).then(frameData => {
+                    if (!(chunkIdx in this._requestedChunks)
+                        || !this._requestedChunks[chunkIdx].requestedFrames.has(requestedFrame)) {
+                        reject(requestedFrame);
+                    }
 
-                if (this._loadCounter <= 0) {
-                    this._stack = [];
-                }
+                    if (frameData !== null) {
+                        this._requestedChunks[chunkIdx].requestedFrames.delete(requestedFrame);
+                        this._requestedChunks[chunkIdx].buffer[requestedFrame] = frameData;
+                        if (this._requestedChunks[chunkIdx].requestedFrames.size === 0) {
+                            const bufferedframes = Object.keys(
+                                this._requestedChunks[chunkIdx].buffer,
+                            ).map(f => +f);
+                            this._requestedChunks[chunkIdx].resolve(new Set(bufferedframes));
+                        }
+                    }
+                }).catch(error => {
+                    reject(error);
+                });
+            }
+        });
+    }
 
-                if (!this._stack.length && this._required == null) {
-                    clearInterval(this._loadInterval);
-                    this._preloadRunned = false;
-                    this._loadInterval = null;
-                    return;
-                }
+    fillBuffer(startFrame, frameStep = 1, count = null) {
+        const freeSize = this.getFreeBufferSize();
```

Comment:
In the future try to avoid any code before ```return new Promise(...```
If you actually need some actions before, do the function async